File: src/lib/objmdl/animation/animation.ts
import assert from 'assert';
import chalk from 'chalk';
import { readFileSync } from 'fs';
import _ from 'lodash';

import { Vector3 } from '@/lib/math/common';

import { BlizzardNull } from '../../constants';
import { SkinWeight } from '../mdl/components/geoset';
import { GlobalSequence } from '../mdl/components/global-sequence';
import { Bone, NodeFlag } from '../mdl/components/node/node';
import { MDL, WowAttachment } from '../mdl/mdl';
import { wowToWc3Interpolation } from '../utils';
import { getWacraftSequenceData, getWowAnimName, isLoopAnimation } from './animation_mapper';
import { getBoneName } from './bones_mapper';

export namespace Data {
  export interface Bone {
    boneID: number
    flags: number
    parentBone: number
    subMeshID: number
    boneNameCRC: number
    translation: Translation
    rotation: Rotation
    scale: Scale
    pivot: [number, number, number]
  }

  export interface Translation {
    globalSeq: number
    interpolation: number
    timestamps: (number[] | null)[]
    values: [number, number, number][][]
  }

  export interface Rotation {
    globalSeq: number
    interpolation: number
    timestamps: (number[] | null)[]
    values: [number, number, number, number][][]
  }

  export interface Scale {
    globalSeq: number
    interpolation: number
    timestamps: (number[] | null)[]
    values: [number, number, number][][]
  }

  export interface Animation {
    id: number
    variationIndex: number
    duration: number
    movespeed: number
    flags: number
    frequency: number
    padding: number
    replayMin: number
    replayMax: number
    blendTimeIn: number
    blendTimeOut: number
    boxPosMin: [number, number, number]
    boxPosMax: [number, number, number]
    boxRadius: number
    variationNext: number
    aliasNext: number
  }

  export interface Attachment {
    id: number
    bone: number
    unknown: number
    position: Vector3
    animateAttached: {
      globalSeq: number
      interpolation: number
      timestamps: number[][]
      values: number[][]
    }
  }
}

export interface AnimationData {
  bones: Data.Bone[]
  animations?: Data.Animation[]
  boneWeights: number[]
  boneIndicies: number[]
}

const debug = false;

export class AnimationFile implements AnimationData {
  bones: Data.Bone[];

  animations?: Data.Animation[];

  boneWeights: number[];

  boneIndicies: number[];

  attachments: Data.Attachment[];

  isLoaded = false;

  constructor(public filePath: string) {
    try {
      console.log("Loading animation file", this.filePath);
      const start = performance.now();
      Object.assign(this, JSON.parse(readFileSync(filePath, 'utf-8')));
      debug && console.log('AnimationFile load took', chalk.yellow(((performance.now() - start) / 1000).toFixed(2)), 's');
      this.isLoaded = true;
    } catch (e) {
      if (e.code === 'ENOENT') {
        // file not exist, do not load.
        return;
      }
      throw e;
    }
  }

  /**
   * `boneIndices[vertexId]` is the list of bones attached to the vertex vertexId
   */
  toMdl(globalSequences: GlobalSequence[]): Pick<MDL, 'bones' | 'sequences'> & {
    skinWeights: SkinWeight[][],
    wowAttachments: WowAttachment[],
  } {
    const start = performance.now();
    if (!this.isLoaded) {
      throw new Error('Animation file is not loaded');
    }

    const excludedAnimation = new Set<number>();

    const globalSequenceMap = new Map<number, GlobalSequence>(globalSequences.map((gs) => [gs.id, gs]));
    function getGlobalSeq(id: number) {
      if (!globalSequenceMap.has(id)) {
        const newGs = {
          id, duration: 1,
        };
        globalSequences.push(newGs);
        globalSequenceMap.set(id, newGs);
      }
      return globalSequenceMap.get(id);
    }

    const bones: MDL['bones'] = this.bones.map((bone, boneI) => {
      const mdlBone: Bone = {
        type: 'Bone',
        name: getBoneName(bone.boneID, boneI, bone.boneNameCRC),
        flags: [],
        translation: {
          interpolation: wowToWc3Interpolation(bone.translation.interpolation),
          globalSeq: bone.translation.globalSeq !== BlizzardNull ? getGlobalSeq(bone.translation.globalSeq) : undefined,
          keyFrames: new Map(),
          type: 'translation',
        },
        rotation: {
          interpolation: wowToWc3Interpolation(bone.rotation.interpolation),
          globalSeq: bone.rotation.globalSeq !== BlizzardNull ? getGlobalSeq(bone.rotation.globalSeq) : undefined,
          keyFrames: new Map(),
          type: 'rotation',
        },
        scaling: {
          interpolation: wowToWc3Interpolation(bone.scale.interpolation),
          globalSeq: bone.scale.globalSeq !== BlizzardNull ? getGlobalSeq(bone.scale.globalSeq) : undefined,
          keyFrames: new Map(),
          type: 'scaling',
        },
        pivotPoint: [bone.pivot[0], -bone.pivot[2], bone.pivot[1]],
      };

      // https://wowdev.wiki/M2#Bones bone flags
      if (bone.flags & 0x1) mdlBone.flags.push(NodeFlag.DONTINHERIT_TRANSLATION);
      if (bone.flags & 0x2) mdlBone.flags.push(NodeFlag.DONTINHERIT_SCALING);
      if (bone.flags & 0x4) mdlBone.flags.push(NodeFlag.DONTINHERIT_ROTATION);
      if (bone.flags & 0x8) mdlBone.flags.push(NodeFlag.BILLBOARDED);
      if (bone.flags & 0x10) mdlBone.flags.push(NodeFlag.BILLBOARD_LOCK_Y); // X and Z are swapped?
      if (bone.flags & 0x20) mdlBone.flags.push(NodeFlag.BILLBOARD_LOCK_X);
      if (bone.flags & 0x40) mdlBone.flags.push(NodeFlag.BILLBOARD_LOCK_Z); // X and Z are swapped?

      const isValidTimestamps = (timestamps: number[] | null, animId: number): timestamps is number[] => {
        const isNull = timestamps == null;
        const isNotIncreasing = timestamps != null
          && timestamps.some((_, i) => i > 0 && timestamps[i] < timestamps[i - 1] || timestamps[i] > 9999999);
        const isInvalid = isNull || isNotIncreasing;
        if (isInvalid) {
          console.warn(`Invalid timestamps ${timestamps} for bone ${bone.boneNameCRC} in animation ${this.animations?.[animId]?.id}`, { isNull, isNotIncreasing});
        }
        return !isInvalid;
      }

      // Translation
      let accumTime = 0;
      bone.translation.timestamps.forEach((timestamps, animId) => {
        const animation = this.animations?.[animId];
        if (animation == null) {
          return;
        }

        const startTime = accumTime;
        accumTime += animation.duration + 1;

        if (excludedAnimation.has(animId)) {
          return;
        }

        if (!isValidTimestamps(timestamps, animId)) {
          if (timestamps != null) {
            excludedAnimation.add(animId);
          }
          return;
        }

        let maxTimestamp = -Infinity;
        timestamps.forEach((timestamp, timestampI) => {
          const values = bone.translation.values[animId][timestampI];
          if (values.length !== 3) {
            throw new Error(`Invalid Vector3 ${values.toString()} for bone ${bone.boneNameCRC} in animation ${this.animations?.[animId]?.id}`);
          }

          const [x, y, z] = values;
          if (x == null || y == null || z == null) return;
          if (Math.abs(x) > 999999 || Math.abs(y) > 999999 || Math.abs(z) > 999999) return;
          mdlBone.translation!.keyFrames.set(timestamp + startTime, [x, -z, y]);
          maxTimestamp = Math.max(maxTimestamp, timestamp + startTime);
        });
        if (maxTimestamp >= -1 && !mdlBone.translation!.globalSeq) {
          mdlBone.translation!.keyFrames.set(startTime + animation.duration, [...mdlBone.translation!.keyFrames.get(maxTimestamp)!]);
        }
      });
      // Rotation
      accumTime = 0;
      bone.rotation.timestamps.forEach((timestamps, animId) => {
        const animation = this.animations?.[animId];
        if (animation == null) {
          return;
        }
        const startTime = accumTime;
        accumTime += animation.duration + 1;

        if (excludedAnimation.has(animId)) {
          return;
        }
        if (!isValidTimestamps(timestamps, animId)) {
          if (timestamps != null) {
            excludedAnimation.add(animId);
          }
          return;
        }
        let maxTimestamp = -Infinity;
        timestamps.forEach((timestamp, timestampI) => {
          const [w, x, y, z] = bone.rotation.values[animId][timestampI];
          if (x == null || y == null || z == null) return;
          if (Math.abs(x) > 999999 || Math.abs(y) > 999999 || Math.abs(z) > 999999) return;
          mdlBone.rotation!.keyFrames.set(timestamp + startTime, [w, -y, x, z]);
          maxTimestamp = Math.max(maxTimestamp, timestamp + startTime);
        });
        if (maxTimestamp >= -1 && !mdlBone.rotation!.globalSeq) {
          mdlBone.rotation!.keyFrames.set(startTime + animation.duration, [...mdlBone.rotation!.keyFrames.get(maxTimestamp)!]);
        }
      });

      // Scaling
      accumTime = 0;
      bone.scale.timestamps.forEach((timestamps, animId) => {
        const animation = this.animations?.[animId];
        if (animation == null) {
          return;
        }
        const startTime = accumTime;
        accumTime += animation.duration + 1;

        if (excludedAnimation.has(animId)) {
          return;
        }
        if (!isValidTimestamps(timestamps, animId)) {
          if (timestamps != null) {
            excludedAnimation.add(animId);
          }
          return;
        }
        let maxTimestamp = -Infinity;
        timestamps.forEach((timestamp, timestampI) => {
          const [x, y, z] = bone.scale.values[animId][timestampI];
          if (x == null || y == null || z == null) return;
          if (Math.abs(x) > 999999 || Math.abs(y) > 999999 || Math.abs(z) > 999999) return;
          mdlBone.scaling!.keyFrames.set(timestamp + startTime, [x, z, y]);
          maxTimestamp = Math.max(maxTimestamp, timestamp + startTime);
        });
        if (maxTimestamp >= -1 && !mdlBone.scaling!.globalSeq) {
          mdlBone.scaling!.keyFrames.set(startTime + animation.duration, [...mdlBone.scaling!.keyFrames.get(maxTimestamp)!]);
        }
      });
      return mdlBone;
    });

    // Fill in parent and safeguard
    bones.forEach((mdlBone, i) => {
      const dataBone = this.bones[i];
      mdlBone.parent = dataBone.parentBone > -1 ? bones[dataBone.parentBone] : undefined;
      if (mdlBone.translation!.keyFrames.size + mdlBone.rotation!.keyFrames.size + mdlBone.scaling!.keyFrames.size === 0 && mdlBone.parent != null) {
        // This bone has parent, it must have some transformation, else WC3 will render crazy
        mdlBone.translation!.keyFrames.set(0, [0, 0, 0]); // no translation
      }
      if (!mdlBone.translation?.keyFrames.size) mdlBone.translation = undefined;
      if (!mdlBone.rotation?.keyFrames.size) mdlBone.rotation = undefined;
      if (!mdlBone.scaling?.keyFrames.size) mdlBone.scaling = undefined;
    });

    const skinWeights: SkinWeight[][] = [];
    for (let i = 0; i < (this.boneIndicies?.length ?? 0); i += 4) {
      const weights = this.boneWeights.slice(i, i + 4);
      assert.ok(_.sum(weights) > 0);
      const indices = this.boneIndicies.slice(i, i + 4);
      const skinWeight: SkinWeight[] = [];
      weights.forEach((w, j) => {
        if (w > 0) {
          skinWeight.push({
            bone: bones[indices[j]],
            weight: weights[j],
          });
        }
      });
      skinWeight.sort((w1, w2) => w2.weight - w1.weight);
      skinWeights.push(skinWeight);
    }

    // Compute `Sequences`
    let seqAccumTime = 0;
    const sequences: MDL['sequences'] = (this.animations ?? []).map((animation, animId) => {
      const wowAnimName = getWowAnimName(animation.id);
      const seqData = getWacraftSequenceData(animation);
      if (excludedAnimation.has(animId)) {
        seqAccumTime += animation.duration + 1;
        // console.log("Skip animation", seqData.wc3Name, "of", this.filePath)
        return null;
      }

      const sequence: typeof sequences[number] = {
        name: seqData.wc3Name,
        data: seqData,
        interval: [seqAccumTime, seqAccumTime + animation.duration],
        moveSpeed: animation.movespeed,
        minimumExtent: [animation.boxPosMin[0], -animation.boxPosMax[2], animation.boxPosMin[1]],
        maximumExtent: [animation.boxPosMax[0], -animation.boxPosMin[2], animation.boxPosMax[1]],
        boundsRadius: animation.boxRadius,
        nonLooping: !isLoopAnimation(wowAnimName),
      };
      seqAccumTime += animation.duration + 1;
      return sequence;
    }).filter((seq) => !!seq);

    // Find secondary Stand animations and set their rarity

    const standSeqs = sequences.filter((seq) => seq.data.wc3Name === 'Stand');
    if (standSeqs.length > 1) {
      const mainStandSeq = standSeqs.reduce((best, seq) => (seq.data.wowFrequency > best.data.wowFrequency ? seq : best));
      standSeqs.filter((seq) => seq !== mainStandSeq).forEach((seq) => seq.rarity = 4);
    }

    if (sequences.length === 0) {
      // Model with sequence will crash Wc3
      sequences.push({
        name: 'Stand',
        data: {
          wowName: '', attackTag: '', wc3Name: 'Stand', wowVariant: 0, wowFrequency: 0,
        },
        interval: [0, 1000],
        moveSpeed: 0,
        minimumExtent: [-1, -1, -1],
        maximumExtent: [1, 1, 1],
        boundsRadius: 1,
        nonLooping: false,
      });
    }

    const wowAttachments = this.extractWowAttachments(bones);
    debug && console.log('AnimationFile toMdl took', chalk.yellow(((performance.now() - start) / 1000).toFixed(2)), 's');

    return {
      sequences,
      bones,
      skinWeights,
      wowAttachments,
    };
  }

  private extractWowAttachments(bones: Bone[]): WowAttachment[] {
    if (!this.isLoaded || !this.attachments) {
      return [];
    }

    const result: WowAttachment[] = [];

    this.attachments.forEach((attachment) => {
      const parent = bones[attachment.bone];
      if (!parent) {
        console.warn(`Attachment ${attachment.id} references non-existent bone ${attachment.bone}`);
        return;
      }

      const [x, y, z] = attachment.position;
      result.push({
        wowAttachmentId: attachment.id,
        bone: parent,
        pivotPoint: [x, -z, y],
      });
    });

    return result;
  }
}

File: golang/internal/converter/wowmodel/bundle/animation/animation.go
package animation

import (
	"fmt"
	"log"
	stdmath "math"
	"sort"

	"github.com/pqhuy98/wow-converter/internal/ansi"
	"github.com/pqhuy98/wow-converter/internal/config"
	animmap "github.com/pqhuy98/wow-converter/internal/converter/wowmodel/animation"
	bundleutils "github.com/pqhuy98/wow-converter/internal/converter/wowmodel/bundle/utils"
	"github.com/pqhuy98/wow-converter/internal/formats/mdl"
	"github.com/pqhuy98/wow-converter/internal/formats/mdl/components"
	imath "github.com/pqhuy98/wow-converter/internal/math"
)

// BoneData is a bone entry from _bones.json equivalent.
type BoneData struct {
	BoneID      int32      `json:"boneID"`
	Flags       uint32     `json:"flags"`
	ParentBone  int16      `json:"parentBone"`
	SubMeshID   uint16     `json:"subMeshID"`
	BoneNameCRC uint32     `json:"boneNameCRC"`
	Translation TrackData  `json:"translation"`
	Rotation    TrackData  `json:"rotation"`
	Scale       TrackData  `json:"scale"`
	Pivot       [3]float64 `json:"pivot"`
}

// TrackData is animation track data.
type TrackData struct {
	GlobalSeq     uint16        `json:"globalSeq"`
	Interpolation uint16        `json:"interpolation"`
	Timestamps    [][]*uint32   `json:"timestamps"`
	Values        [][][]float64 `json:"values"`
}

// AnimationData is animation metadata.
type AnimationData struct {
	ID             uint16     `json:"id"`
	VariationIndex uint16     `json:"variationIndex"`
	Duration       uint32     `json:"duration"`
	MoveSpeed      float32    `json:"movespeed"`
	Flags          uint32     `json:"flags"`
	Frequency      uint32     `json:"frequency"`
	ReplayMin      uint32     `json:"replayMin"`
	ReplayMax      uint32     `json:"replayMax"`
	BlendTimeIn    uint16     `json:"blendTimeIn"`
	BlendTimeOut   uint16     `json:"blendTimeOut"`
	BoxPosMin      [3]float64 `json:"boxPosMin"`
	BoxPosMax      [3]float64 `json:"boxPosMax"`
	BoxRadius      float32    `json:"boxRadius"`
	VariationNext  int16      `json:"variationNext"`
	AliasNext      int16      `json:"aliasNext"`
}

// AttachmentData is a WoW attachment point.
type AttachmentData struct {
	ID       uint32     `json:"id"`
	Bone     uint16     `json:"bone"`
	Unknown  uint16     `json:"unknown"`
	Position [3]float64 `json:"position"`
}

// File holds parsed bones/animation data.
type File struct {
	FilePath    string
	Config      config.Config
	Bones       []BoneData
	Animations  []AnimationData
	BoneWeights []float64
	BoneIndices []float64
	Attachments []AttachmentData
	IsLoaded    bool
}

// NewFile creates an animation file placeholder.
func NewFile(filePath string, cfg config.Config) *File {
	return &File{FilePath: filePath, Config: cfg}
}

// LoadFromData populates from direct M2 pipeline output.
func (f *File) LoadFromData(data map[string]any) {
	if v, ok := data["bones"].([]any); ok {
		f.Bones = parseBones(v)
	}
	if v, ok := data["animations"].([]any); ok {
		f.Animations = parseAnimations(v)
	}
	if v, ok := data["boneWeights"].([]any); ok {
		f.BoneWeights = toFloatSlice(v)
	}
	if v, ok := data["boneIndicies"].([]any); ok {
		f.BoneIndices = toFloatSlice(v)
	}
	if v, ok := data["attachments"].([]any); ok {
		f.Attachments = parseAttachments(v)
	}
	f.IsLoaded = true
}

// ToMdlResult is the MDL animation conversion output.
type ToMdlResult struct {
	Bones          []*components.Bone
	Sequences      []components.Sequence
	SkinWeights    [][]components.SkinWeight
	WowAttachments []components.WowAttachment
}

// ToMdl converts bones data to MDL nodes.
func (f *File) ToMdl(globalSequences *[]*components.GlobalSequence) ToMdlResult {
	if !f.IsLoaded {
		panic("animation file is not loaded")
	}
	excluded := map[int]struct{}{}
	gsMap := map[int]*components.GlobalSequence{}
	for _, gs := range *globalSequences {
		if gs == nil {
			continue
		}
		key := gs.ID
		if gs.HasRawID {
			key = gs.RawID
		}
		gsMap[key] = gs
	}
	getGS := func(id int) *components.GlobalSequence {
		if id == config.BlizzardNull {
			return nil
		}
		if gs, ok := gsMap[id]; ok {
			return gs
		}
		created := components.NewGlobalSequence(id, 1)
		ptr := &created
		*globalSequences = append(*globalSequences, ptr)
		key := ptr.ID
		if ptr.HasRawID {
			key = ptr.RawID
		}
		gsMap[key] = ptr
		return ptr
	}

	bones := make([]*components.Bone, len(f.Bones))
	for i, bone := range f.Bones {
		mdlBone := &components.Bone{
			NodeBase: components.NodeBase{
				Type:       "Bone",
				Name:       animmap.GetBoneName(int(bone.BoneID), i, bone.BoneNameCRC),
				PivotPoint: imath.Vector3{bone.Pivot[0], -bone.Pivot[2], bone.Pivot[1]},
				Translation: &components.Animation{
					Type:          components.AnimTypeTranslation,
					Interpolation: bundleutils.WowToWc3Interpolation(bone.Translation.Interpolation),
					KeyFrames:     map[int]any{},
				},
				Rotation: &components.Animation{
					Type:          components.AnimTypeRotation,
					Interpolation: bundleutils.WowToWc3Interpolation(bone.Rotation.Interpolation),
					KeyFrames:     map[int]any{},
				},
				Scaling: &components.Animation{
					Type:          components.AnimTypeScaling,
					Interpolation: bundleutils.WowToWc3Interpolation(bone.Scale.Interpolation),
					KeyFrames:     map[int]any{},
				},
			},
		}
		if int(bone.Translation.GlobalSeq) != config.BlizzardNull {
			mdlBone.Translation.GlobalSeq = getGS(int(bone.Translation.GlobalSeq))
		}
		if int(bone.Rotation.GlobalSeq) != config.BlizzardNull {
			mdlBone.Rotation.GlobalSeq = getGS(int(bone.Rotation.GlobalSeq))
		}
		if int(bone.Scale.GlobalSeq) != config.BlizzardNull {
			mdlBone.Scaling.GlobalSeq = getGS(int(bone.Scale.GlobalSeq))
		}
		if bone.Flags&0x1 != 0 {
			mdlBone.Flags = append(mdlBone.Flags, components.NodeFlagDontInheritTranslation)
		}
		if bone.Flags&0x2 != 0 {
			mdlBone.Flags = append(mdlBone.Flags, components.NodeFlagDontInheritScaling)
		}
		if bone.Flags&0x4 != 0 {
			mdlBone.Flags = append(mdlBone.Flags, components.NodeFlagDontInheritRotation)
		}
		if bone.Flags&0x8 != 0 {
			mdlBone.Flags = append(mdlBone.Flags, components.NodeFlagBillboarded)
		}
		if bone.Flags&0x10 != 0 {
			mdlBone.Flags = append(mdlBone.Flags, components.NodeFlagBillboardLockY)
		}
		if bone.Flags&0x20 != 0 {
			mdlBone.Flags = append(mdlBone.Flags, components.NodeFlagBillboardLockX)
		}
		if bone.Flags&0x40 != 0 {
			mdlBone.Flags = append(mdlBone.Flags, components.NodeFlagBillboardLockZ)
		}
		applyTrack(mdlBone.Translation, bone.Translation, f.Animations, excluded, func(v []float64) any {
			return imath.Vector3{v[0], -v[2], v[1]}
		})
		applyRotTrack(mdlBone.Rotation, bone.Rotation, f.Animations, excluded)
		applyTrack(mdlBone.Scaling, bone.Scale, f.Animations, excluded, func(v []float64) any {
			return imath.Vector3{v[0], v[2], v[1]}
		})
		bones[i] = mdlBone
	}

	for i, mdlBone := range bones {
		parent := int(f.Bones[i].ParentBone)
		if parent > -1 && parent < len(bones) {
			mdlBone.Parent = bones[parent]
		}
		if countKeyframes(mdlBone) == 0 && mdlBone.Parent != nil {
			mdlBone.Translation.KeyFrames[0] = imath.Vector3{}
		}
		if len(mdlBone.Translation.KeyFrames) == 0 {
			mdlBone.Translation = nil
		}
		if len(mdlBone.Rotation.KeyFrames) == 0 {
			mdlBone.Rotation = nil
		}
		if len(mdlBone.Scaling.KeyFrames) == 0 {
			mdlBone.Scaling = nil
		}
	}

	skinWeights := buildSkinWeights(f, bones)

	var sequences []components.Sequence
	seqAccum := 0
	for animID, anim := range f.Animations {
		meta := animmap.AnimationMeta{
			ID: anim.ID, VariationIndex: anim.VariationIndex, Duration: anim.Duration,
			MoveSpeed: anim.MoveSpeed, Frequency: anim.Frequency,
			BoxPosMin: [3]float32{float32(anim.BoxPosMin[0]), float32(anim.BoxPosMin[1]), float32(anim.BoxPosMin[2])},
			BoxPosMax: [3]float32{float32(anim.BoxPosMax[0]), float32(anim.BoxPosMax[1]), float32(anim.BoxPosMax[2])},
			BoxRadius: anim.BoxRadius,
		}
		seqData := animmap.GetWarcraftSequenceData(meta)
		if _, skip := excluded[animID]; skip {
			seqAccum += int(anim.Duration) + 1
			log.Printf("%s", ansi.Redf("Skip unreadable animation %s of %s", seqData.WowName, f.FilePath))
			continue
		}
		wowName := animmap.GetWowAnimName(int(anim.ID))
		seq := components.Sequence{
			Name:      seqData.WC3Name,
			Data:      seqData,
			Interval:  [2]int{seqAccum, seqAccum + int(anim.Duration)},
			MoveSpeed: float64(anim.MoveSpeed),
			Bound: components.Bound{
				MinimumExtent: imath.Vector3{anim.BoxPosMin[0], -float64(anim.BoxPosMax[2]), anim.BoxPosMin[1]},
				MaximumExtent: imath.Vector3{anim.BoxPosMax[0], -float64(anim.BoxPosMin[2]), anim.BoxPosMax[1]},
				BoundsRadius:  float64(anim.BoxRadius),
			},
			NonLooping: !animmap.IsLoopAnimation(wowName),
		}
		sequences = append(sequences, seq)
		seqAccum += int(anim.Duration) + 1
	}

	wowAttachments := extractWowAttachments(f, bones)

	standLikeNames := []string{"Stand", "Stand Alternate", "Walk", "Walk Alternate"}
	for _, baseName := range standLikeNames {
		var matches []*components.Sequence
		for i := range sequences {
			if sequences[i].Data.WC3Name == baseName {
				matches = append(matches, &sequences[i])
			}
		}
		if len(matches) <= 1 {
			continue
		}
		best := matches[0]
		for _, seq := range matches[1:] {
			if seq.Data.WowFrequency > best.Data.WowFrequency {
				best = seq
			}
		}
		rarity := 4
		for _, seq := range matches {
			if seq != best {
				seq.Rarity = &rarity
			}
		}
	}

	return ToMdlResult{Bones: bones, Sequences: sequences, SkinWeights: skinWeights, WowAttachments: wowAttachments}
}

func countKeyframes(b *components.Bone) int {
	n := 0
	if b.Translation != nil {
		n += len(b.Translation.KeyFrames)
	}
	if b.Rotation != nil {
		n += len(b.Rotation.KeyFrames)
	}
	if b.Scaling != nil {
		n += len(b.Scaling.KeyFrames)
	}
	return n
}

func applyTrack(anim *components.Animation, track TrackData, animations []AnimationData, excluded map[int]struct{}, conv func([]float64) any) {
	accum := 0
	for animID, timestamps := range track.Timestamps {
		if animID >= len(animations) {
			continue
		}
		animation := animations[animID]
		start := accum
		accum += int(animation.Duration) + 1
		if _, skip := excluded[animID]; skip {
			continue
		}
		if !validTimestamps(timestamps) {
			if timestamps != nil {
				excluded[animID] = struct{}{}
			}
			continue
		}
		maxTS := -1
		for ti, ts := range timestamps {
			if ts == nil {
				continue
			}
			t := int(*ts) + start
			if animID >= len(track.Values) || ti >= len(track.Values[animID]) {
				continue
			}
			vals := track.Values[animID][ti]
			if len(vals) < 3 {
				continue
			}
			if mathAbs(vals[0]) > 999999 || mathAbs(vals[1]) > 999999 || mathAbs(vals[2]) > 999999 {
				continue
			}
			anim.KeyFrames[t] = conv(vals)
			if t > maxTS {
				maxTS = t
			}
		}
		if maxTS >= -1 && anim.GlobalSeq == nil && maxTS >= 0 {
			if v, ok := anim.KeyFrames[maxTS]; ok {
				anim.KeyFrames[start+int(animation.Duration)] = v
			}
		}
	}
}

func applyRotTrack(anim *components.Animation, track TrackData, animations []AnimationData, excluded map[int]struct{}) {
	accum := 0
	for animID, timestamps := range track.Timestamps {
		if animID >= len(animations) {
			continue
		}
		animation := animations[animID]
		start := accum
		accum += int(animation.Duration) + 1
		if _, skip := excluded[animID]; skip {
			continue
		}
		if !validTimestamps(timestamps) {
			if timestamps != nil {
				excluded[animID] = struct{}{}
			}
			continue
		}
		maxTS := -1
		for ti, ts := range timestamps {
			if ts == nil {
				continue
			}
			t := int(*ts) + start
			if animID >= len(track.Values) || ti >= len(track.Values[animID]) {
				continue
			}
			vals := track.Values[animID][ti]
			if len(vals) < 4 {
				continue
			}
			anim.KeyFrames[t] = imath.QuaternionRotation{vals[0], -vals[2], vals[1], vals[3]}
			if t > maxTS {
				maxTS = t
			}
		}
		if maxTS >= -1 && anim.GlobalSeq == nil && maxTS >= 0 {
			if v, ok := anim.KeyFrames[maxTS]; ok {
				anim.KeyFrames[start+int(animation.Duration)] = v
			}
		}
	}
}

func validTimestamps(timestamps []*uint32) bool {
	if timestamps == nil {
		return false
	}
	var prev *uint32
	for _, ts := range timestamps {
		if ts == nil {
			continue
		}
		if *ts > 9999999 {
			return false
		}
		if prev != nil && *ts < *prev {
			return false
		}
		prev = ts
	}
	return true
}

func buildSkinWeights(f *File, bones []*components.Bone) [][]components.SkinWeight {
	var out [][]components.SkinWeight
	for i := 0; i+3 < len(f.BoneIndices); i += 4 {
		weights := f.BoneWeights[i : i+4]
		indices := f.BoneIndices[i : i+4]
		var sw []components.SkinWeight
		sum := 0.0
		for _, w := range weights {
			sum += w
		}
		if sum <= 0 {
			continue
		}
		for j, w := range weights {
			if w > 0 {
				idx := int(indices[j])
				if idx < len(bones) {
					sw = append(sw, components.SkinWeight{Bone: bones[idx], Weight: int(w)})
				}
			}
		}
		sort.Slice(sw, func(a, b int) bool { return sw[a].Weight > sw[b].Weight })
		out = append(out, sw)
	}
	return out
}

func extractWowAttachments(f *File, bones []*components.Bone) []components.WowAttachment {
	var result []components.WowAttachment
	for _, att := range f.Attachments {
		if int(att.Bone) >= len(bones) {
			continue
		}
		pos := att.Position
		result = append(result, components.WowAttachment{
			WowAttachmentID: int(att.ID),
			Bone:            bones[att.Bone],
			PivotPoint:      imath.Vector3{float64(pos[0]), -float64(pos[2]), float64(pos[1])},
		})
	}
	return result
}

func mathAbs(v float64) float64 {
	if v < 0 {
		return -v
	}
	return v
}

func toFloatSlice(v []any) []float64 {
	out := make([]float64, len(v))
	for i, x := range v {
		switch n := x.(type) {
		case float64:
			out[i] = n
		case int:
			out[i] = float64(n)
		}
	}
	return out
}

func parseBones(v []any) []BoneData {
	// Bones arrive pre-normalized from buildBonesData via json round-trip in direct pipeline.
	return nil
}

func parseAnimations(v []any) []AnimationData { return nil }

func parseAttachments(v []any) []AttachmentData { return nil }

// LoadFromBonesData loads typed bones data directly.
func (f *File) LoadFromBonesData(bones []BoneData, animations []AnimationData, weights, indices []float64, attachments []AttachmentData) {
	f.Bones = bones
	f.Animations = animations
	f.BoneWeights = weights
	f.BoneIndices = indices
	f.Attachments = attachments
	f.IsLoaded = true
}

// Ensure mdl import is used for wow attachment type check
var _ = mdl.MDL{}

func init() {
	_ = fmt.Sprintf
	_ = stdmath.MaxFloat64
}
